clippy_lints/src/empty_with_brackets.rs

```diff
@@ -97,22 +113,63 @@ impl EarlyLintPass for EmptyWithBrackets {
         }
     }
 
-    fn check_variant(&mut self, cx: &EarlyContext<'_>, variant: &Variant) {
+    fn check_variant(&mut self, cx: &LateContext<'_>, variant: &Variant<'_>) {
+        // Don't lint pub enums
+        if cx.effective_visibilities.is_reachable(variant.def_id) {
+            return;
+        }
+
         let span_after_ident = variant.span.with_lo(variant.ident.span.hi());
 
-        if has_brackets(&variant.data) && has_no_fields(cx, &variant.data, span_after_ident) {
+        if has_no_fields(cx, &variant.data, span_after_ident) {
+            match variant.data {
+                VariantData::Struct { .. } => {
+                    span_lint_and_then(
+                        cx,
+                        EMPTY_ENUM_VARIANTS_WITH_BRACKETS,
+                        span_after_ident,
+                        "enum variant has empty brackets",
+                        |diagnostic| {
+                            diagnostic.span_suggestion_hidden(
+                                span_after_ident,
+                                "remove the brackets",
+                                "",
+                                Applicability::MaybeIncorrect,
+                            );
+                        },
+                    );
+                },
+                VariantData::Tuple(..) => {
+                    if let Some(local_def_id) = variant.data.ctor_def_id()
+                        && !self.empty_tuple_enum_variants.contains_key(&local_def_id.to_def_id())
+                    {
+                        self.empty_tuple_enum_variants
+                            .insert(local_def_id.to_def_id(), Usage::Unused(span_after_ident));
+                    }
+                },
+                VariantData::Unit(..) => {},
+            }
+        }
+    }
+
+    fn check_expr(&mut self, _cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if let Some(def_id) = check_expr_for_enum_as_function(expr) {
```

Comment:
Also, this should probably check that the parent expression of the path is not a function call to this expression.

It still seems worth emitting a warning on e.g. `X::V();` as that can just be replaced with `X::V;`.

Reply:
Isn't this lint limited to checking the use of redundant brackets in the definition of variants?

Reply:
Yes, but this PR changes the lint to not emit a warning on them if there's any `Enum::Variant();` expression because the callee is a path to the constructor and gets inserted as `Used`, right?

Reply:
How do I check if the constructor is just called immediately (like `X::V();`)?

Reply:
You can go from `X::V` to `X::V()` with `tcx.parent_hir_node(path_expr.hir_id)`.

It'll be a `Node::Expr(ExprKind::Call(callee))` with `callee.hir_id == path_expr.hir_id` if the path is the callee of a function call.

Something like
```rs
fn is_callee(tcx: TyCtxt<'_>, expr: &Expr<'_>) -> bool {
  if let Some(Node::Expr(parent)) = tcx.parent_hir_node(expr.hir_id)
    && let ExprKind::Call(callee, ..) = parent.kind
  {
    callee.hir_id == expr.hir_id
  } else {
    false
  }
}

// if the path is the callee of a function call, it's not really used as a value and we can still warn it
if let Some(def_id) = check_expr_for_enum_as_function(expr) && !is_callee(cx.tcx, expr) {
  // ... add Used to the map ...
```

> And what would the ideal lint message and suggestion be in such cases?

I don't think we need to emit a "special" warning for this (if I understand this correctly). If we ignore fn calls here and don't insert `Used` in those cases, then the enum variant will naturally get picked up as `Unused` by the other part of this lint's logic and continue to emit the normal warning.

Reply:
Re-reading my initial comment here I can see where the confusion probably came from.

I said "worth emitting a warning on e.g. `X::V()`" but what I really meant was that we can emit a warning on `enum X { V() }` if there is a call expression `X::V();`, sorry for that.

Reply:
Won't that be a `suggestion-causes-error` issue?

Reply:
Good point. This has already been an issue with this lint before that the suggestion can cause errors at use sites though, but I don't think it's a problem given that the applicability is not `MachineApplicable` and so tools like rustfix won't automatically apply the suggestion and already implies that it requires some additional manual work.

We could just change the help message to include "and adjust the use sites" (or something).

(There'd also be the option to have `enum Usage { Used { replacements: Vec<(Span, String)> } }` that we could pass to `multipart_suggestion` but I don't know how to feel about that, since it could end up as a giant diagnostic with replacements across lots of files, and I don't think we want that)

Reply:
I think we can use `multipart_suggestion` but only point to one use site each time. That way the diagnostic won't be very long and it will also point the programmer in the right direction.

Reply:
How do I check for `X::V();`?

Reply:
I think this should work: https://github.com/rust-lang/rust-clippy/pull/12971#discussion_r1661329677

That is, extending that `if` on L154 (basically where this review comment is pointing at) with another condition that makes sure that `tcx.parent_hir_node(expr.hir_id)` doesnt return `Node::Expr(ExprKind::Call(callee,..))` where callee.hir_id == expr.hir_id

Or did you mean something different?